rados/command.go

```diff
@@ -20,42 +26,23 @@ func (c *Conn) MonCommandWithInputBuffer(args, inputBuffer []byte) ([]byte, stri
 }
 
 func (c *Conn) monCommand(args, inputBuffer []byte) ([]byte, string, error) {
-	argv := C.CString(string(args))
-	defer C.free(unsafe.Pointer(argv))
-
-	var (
-		info               string
-		buffer             []byte
-		outs, outbuf       *C.char
-		outslen, outbuflen C.size_t
-	)
-	inbuf := C.CString(string(inputBuffer))
-	inbufLen := len(inputBuffer)
-	defer C.free(unsafe.Pointer(inbuf))
-
-	ret := C.rados_mon_command(c.cluster,
-		&argv, 1,
-		inbuf,              // bulk input (e.g. crush map)
-		C.size_t(inbufLen), // length inbuf
-		&outbuf,            // buffer
-		&outbuflen,         // buffer length
-		&outs,              // status string
-		&outslen)
-
-	if outslen > 0 {
-		info = C.GoStringN(outs, C.int(outslen))
-		C.free(unsafe.Pointer(outs))
-	}
-	if outbuflen > 0 {
-		buffer = C.GoBytes(unsafe.Pointer(outbuf), C.int(outbuflen))
-		C.free(unsafe.Pointer(outbuf))
-	}
-	if ret != 0 {
-		err := getError(ret)
-		return nil, info, err
-	}
-
-	return buffer, info, nil
+	ci := cutil.NewCommandInput([][]byte{args}, inputBuffer)
+	defer ci.Free()
+	co := cutil.NewCommandOutput(radosBufferFree)
+	defer co.Free()
+
+	ret := C.rados_mon_command(
+		c.cluster,
+		(**C.char)(ci.Cmd()),
```

Comment:
what is the reason to not have `ci.Cmd()` and the like return the `C.*` native types? It would remove the casting in these functions.

Reply:
You can't - cgo only "exports" unexportable lower-cased names. There's a very interesting discussion in a [golang issue on this topic](https://github.com/golang/go/issues/13467). Based on that discussion and a little experimentation I ended up exporting types defined on unsafe.Pointer, which do need to be type-casted as above.
I can experiment some more in this area but this what I settled on for the time being.

Reply:
FWIW I also meant to link to that discussion when I posted this PR but apparently forgot. :-\

Reply:
I think I've hit that restriction before too, so I don't mind the extra casting much. It is not blocker for me.